### ISSUE TITLE
Change Request: Enable `reportUnusedDisableDirectives` config option by default

### ISSUE THREAD

**Issue by bmish** [eslint/eslint#15466]
### 2022-12 UPDATE

I have opened an RFC for this:
* https://github.com/eslint/rfcs/pull/100

### ESLint version

8.5.0

### What problem do you want to solve?

Outdated/unused ESLint disable directive comments (e.g. `notConsole(); // eslint-disable-line no-console`) are currently a blind spot for most apps/repositories. These leftover comments cause clutter and confusion, and detecting them requires going out of one's way by using one of the following existing solutions:

1. Run linting with `eslint --report-unused-disable-directives` (errors, but only applies during CLI)
2. Add `reportUnusedDisableDirectives: true` to .eslintrc.js config file (warns)
3. Enabling third-party rule [eslint-comments/no-unused-disable](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-unused-disable.html) (errors, but requires third-party plugin)

These existing solutions are fine but they lack discoverability and require manual work to adopt in every single repository so are thus not commonly used.
 

### What do you think is the correct solution?

As a breaking change, I propose changing the default value for the config option `reportUnusedDisableDirectives` from false to true. This will cause users to see additional warnings if they have unused disable directives.

Displaying these warnings should help raise awareness about the problem of unused disable directives, and encourage many users to remove unused disable directives caused by their changes.

Note that since these are just warnings, users could also choose to ignore them and let them pile up, as a potential downside. But luckily, the new functionality from running `eslint --fix --report-unused-disable-directives` could be used to fix/remove all of them at any time.

Another note is that we need to limit any default behavior to warnings and not errors for this due to our semver policy. More context in:
* https://github.com/eslint/eslint/pull/14699#discussion_r650863268
  > ...because of our semver policy, which says that bug fixes that produce fewer errors should not break builds.
* https://github.com/eslint/eslint/issues/12703#issuecomment-568582014
  > If we allowed for errors to be reported with reportUnusedDisableDirectives, we would limit what kind of changes we could publish in a semver-patch release (since fixing a bug that would result in fewer errors could create additional unused disable directives).

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
This is an interesting idea. I think there’s a larger question of whether or not we should make the functionality on by default and then not allow it to be turned off? As in, should we remove the command line flag completely?

Worth noting from the docs:

> Warning: When using this option, it is possible that new errors will start being reported whenever ESLint or custom rules are upgraded. For example, suppose a rule has a bug that causes it to report a false positive, and an eslint-disable comment is added to suppress the incorrect report. If the bug is then fixed in a patch release of ESLint, the eslint-disable comment will become unused since ESLint is no longer generating an incorrect report. This will result in a new reported error for the unused directive if the report-unused-disable-directives option is used.

**Comment by bmish** [comment]
@nzakas my proposal is currently only referring to changing the default of the `reportUnusedDisableDirectives` config option, and not the `--report-unused-disable-directives` CLI option. I limited it to the config option which warns instead of errors to avoid affecting the semver policy / issue you quoted.

However, I am open to being more aggressive here. Personally, the behavior I would prefer is to always error on unused directives (regardless of where ESLint is run from).
* If some directives become unused due to bug fixes in ESLint/plugin minor version updates and cause linting to fail, I'm fine with that, especially since they can be easily fixed with `eslint --fix --report-unused-disable-directives`
* "Always erroring on unused directives" is a much simpler story to tell to users than the confusing mix of opt-in warning/error config/CLI options we have today

So if we can consider this more aggressive behavior, let me know, and I'll re-focus the proposal around this.

Removing the one-off CLI/config options for reporting disable directives would be a nice simplification too. Personally, I'd likely be fine removing the options, but I'd be curious if others have a strong use case for disabling/silencing unused disable directive violations.

**Comment by ljharb** [comment]
Related, it would also be SUPER helpful to have a way to report on unused config overrides :-D like, if a shared config defines a rule, and a local config defines it in the same way - or, disables it when it wasn't enabled in the first place - i'd love to know i can remove the line from the local config.

**Comment by nzakas** [comment]
@ljharb thats pretty tangential to the topic at hand. Feel free to open a separate issue if you’d like to discuss that.

@bmish Lets see what other team members think. I’m happy to consider both subtle and aggressive changes. What I don’t want to do is a bunch of incremental changes if we ultimately want to rethink this functionality as a whole.

**Comment by ljharb** [comment]
Sorry about that :-) filed #15476.

**Comment by mdjermanovic** [comment]
> I'd likely be fine removing the options, but I'd be curious if others have a strong use case for disabling/silencing unused disable directive violations.

A use case where reporting unused disable directives is undesirable is when different ESLint configurations are used at different stages of a workflow. In that case, a directive can be unused if the rule wasn't enabled in one config, but may become used with another config at a later stage. We already don't support this use case well, but enforcing the removal of unused disable directives without an opt-out could add more problems for users with such workflows.

**Comment by nzakas** [comment]
Fair enough. What do you think of the proposal as @bmish originally proposed?

> my proposal is currently only referring to changing the default of the reportUnusedDisableDirectives config option, and not the --report-unused-disable-directives CLI option. I limited it to the config option which warns instead of errors to avoid affecting the semver policy / issue you quoted.

I’m not sure I understand how we can change this in one spot but not the other. Can you explain more?

**Comment by mdjermanovic** [comment]
I support changing the default value of the `reportUnusedDisableDirectives` config option to `true` because it would raise awareness of the feature. If for any reason it isn't desirable to report unused directives, user can restore the current behavior by setting `reportUnusedDisableDirectives: false`.

**Comment by mdjermanovic** [comment]
> how we can change this in one spot but not the other

API option `reportUnusedDisableDirectives`, if set, always overrides config option `reportUnusedDisableDirectives`:

https://github.com/eslint/eslint/blob/648fe1adfc02900ee3b96e50043a768a32771fc3/lib/linter/linter.js#L596-L605

The CLI flag `--report-unused-disable-directives` maps to `"error"` on API:

https://github.com/eslint/eslint/blob/648fe1adfc02900ee3b96e50043a768a32771fc3/lib/cli.js#L120

All of this is a bit confusing, but in the end, for CLI users it normalizes to one of these three behaviors:

1. Don't report (neither `reportUnusedDisableDirectives: true` in config nor `--report-unused-disable-directives` on CLI were set)
2. Report as warnings (`reportUnusedDisableDirectives: true` is set in the config, but `--report-unused-disable-directives` wasn't provided on CLI).
3. Report as errors (`--report-unused-disable-directives` was provided on CLI; in this case, the config option is ignored even if it is `reportUnusedDisableDirectives: false`).

The current default behavior is 1. After this change, the default behavior will be 2.

This change wouldn't affect:

* Users who already have `reportUnusedDisableDirectives` config option (either `true` or `false`).
* CLI users with `--report-unused-disable-directives` flag.
* API users with a non-nullish value passed as `reportUnusedDisableDirectives` (`"off"`, `"warn"`, or `"error"`).

**Comment by bmish** [comment]
@mdjermanovic thanks for the context, that's helpful.

It sounds like there is support for my original proposal (i.e. warning on unused directives).

But I haven't yet heard opinions on the more aggressive version which is erroring by default on unused directives. Is erroring still something we should consider now? Is anyone explicitly against erroring? Should we defer this question and just start with warning in ESLint v9 but consider erroring in a subsequent major version? Although as @nzakas said it would be nice to figure out the endgame now.

**Comment by ljharb** [comment]
Since I can't make it error via my shared config, and I don't want to update 300+ projects to pass the command line flag, I'd love the default to be changed.

**Comment by bmish** [comment]
@ljharb that's a great argument for erroring by default, which I support.

I think the main question is just whether we're okay with the potential semver implications of erroring by default. I'm bringing this up because it sounded like in the past we were not okay with them, so I'm wondering if we're open to reconsidering this:
* https://github.com/eslint/eslint/pull/14699#discussion_r650863268
  > ...because of our semver policy, which says that bug fixes that produce fewer errors should not break builds.
* https://github.com/eslint/eslint/issues/12703#issuecomment-568582014
  > If we allowed for errors to be reported with reportUnusedDisableDirectives, we would limit what kind of changes we could publish in a semver-patch release (since fixing a bug that would result in fewer errors could create additional unused disable directives).

**Comment by nzakas** [comment]
Just to make sure I’m understanding correctly, what would the CLI `—report-unused-disable-directives` flag do now? It would just bump up the severity from warning to error?

And that means you can’t opt-out of warning on disable directives from the CLI, only in config? (If true, not a fan.)

@bmish regarding erroring, can you explain how that would interplay with the command line flag?

**Comment by bmish** [comment]
@nzakas your questions have got me thinking that the current boolean design of these options is particularly confusing and unintuitive, making it hard to predict the interplay.

What if I described a new "explicit severity" design:
* The config and CLI options are both changed to accept standard severity values: `off`, `warn`, `error`
* If both options are present, the CLI option takes precedence over (overrides) the config option
* If only one of the options is present, the value of the present option is used
* If neither option is present, the default value is used
* Note: There's only one underlying setting, and the config and CLI options are just two different ways of controlling it

Benefits of this design:
* The user should always be able to achieve the desired effect (`off` / `warn` / `error`), regardless of where they set the option from
* No confusing inconsistency where one of the options triggers errors but the other triggers warnings
* `reportUnusedDisableDirectives: 'error'` can now be part of a shareable config

On top of this design, we can still debate what the default value should be:

* Today: Default is `off`
* Conservative proposal: Default changed to `warn`
* Aggressive proposal: Default changed to `error` (potential semver implication, but this is my preference if acceptable)

**Comment by nzakas** [comment]
That sounds like a good direction to head in.

**Comment by nzakas** [comment]
Assigning to @bmish to continue work

**Comment by kachkaev** [comment]
I’d love to be able to set `reportUnusedDisableDirectives: "error"` instead of `true`. At the moment, it is necessary to add `--report-unused-disable-directives` to every monorepo workspace (example: https://github.com/blockprotocol/blockprotocol/pull/812).

I create a PR that supports `reportUnusedDisableDirectives: "error"` and `reportUnusedDisableDirectives: "warn"`. Any reasons why this would be a bad idea?

**Comment by bmish** [comment]
This will require an RFC. I can work on putting together an RFC based on what I proposed in my previous comment.

**Comment by bmish** [comment]
I have opened an RFC for this:
* https://github.com/eslint/rfcs/pull/100

**Comment by sam3k** [comment]
2023-11-30 tsc-meeting note: waiting on @bmish to finish [PR (draft)](https://github.com/eslint/eslint/pull/17212).

**Comment by bmish** [comment]
Part 1 is done but I'm going to reopen this until part 2 of the RFC is done:

> Part 1: add support for the new severity levels https://github.com/eslint/eslint/pull/17212
> Part 2: `warn` by default

**Comment by mdjermanovic** [comment]
@bmish would you like to submit a PR for part 2 as well? This feature is planned for the first v9 alpha release, which will be next week.

**Comment by bmish** [comment]
Yes I started it in #17879.
